ergotree-ir/src/sigma_protocol/sigma_boolean.rs

```diff
@@ -316,53 +315,24 @@ mod arbitrary {
         }
     }
 
-    fn primitive_type_value() -> BoxedStrategy<SigmaBoolean> {
+    pub fn primitive_type_value() -> BoxedStrategy<SigmaBoolean> {
         prop_oneof![
             any::<ProveDlog>().prop_map_into(),
             any::<ProveDhTuple>().prop_map_into(),
         ]
         .boxed()
     }
 
-    impl Default for ArbSigmaBooleanParams {
-        fn default() -> Self {
-            ArbSigmaBooleanParams::SigmaConjecture(1)
-        }
-    }
-
-    /// Parameters for arbitrary SigmaBoolean generation
-    #[derive(PartialEq, Eq, Debug, Clone, From, TryInto)]
-    pub enum ArbSigmaBooleanParams {
-        /// Sigma proposition
-        ProofOfKnowledge,
-        /// Conjunctions for sigma propositions
-        SigmaConjecture(u8),
-    }
-
     impl Arbitrary for SigmaBoolean {
-        type Parameters = ArbSigmaBooleanParams;
+        type Parameters = ();
         type Strategy = BoxedStrategy<Self>;
 
-        fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
-            match args {
-                ArbSigmaBooleanParams::ProofOfKnowledge => primitive_type_value(),
-                ArbSigmaBooleanParams::SigmaConjecture(depth) => primitive_type_value()
-                    .prop_recursive(depth as u32, 1, 1, |elem| {
-                        prop_oneof![
-                            vec(elem.clone(), 2..5).prop_map(|items| {
-                                SigmaBoolean::SigmaConjecture(SigmaConjecture::Cand(Cand::from(
-                                    items,
-                                )))
-                            }),
-                            vec(elem, 2..5).prop_map(|items| {
-                                SigmaBoolean::SigmaConjecture(SigmaConjecture::Cor(Cor::from(
-                                    items,
-                                )))
-                            }),
-                        ]
-                    })
-                    .boxed(),
-            }
+        fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+            primitive_type_value()
+                .prop_recursive(2, 16, 8, |_elem| {
+                    prop_oneof![any::<Cand>().prop_map_into(), any::<Cor>().prop_map_into(),]
```

Comment:
You don't need to call `any<Cand>()` here. `_elem` is a generator (`Strategy`) which should be used to produce a few values (via `prop::collection::vec(elem, 2..5)`) with something like this:
```rust
prop_oneof![
    prop::collection::vec(elem, 2..=4)
        .prop_map(|elems| Cand { items: elems.try_into().unwrap() }),
    prop::collection::vec(elem, 2..=4)
        .prop_map(|elems| Cor { items: elems.try_into().unwrap() })
]
```

By calling `any<Cand>()` here you're ignoring "recursive" part of `prop_recursive` producing "generic" `Cor` and `Cand`. 

This is also the reason why you cannot call `any<SigmaBoolean>()` in `Arbitrary` impl for `Cor`/`Cand`. Because in the end, it tries to call "itself".

Reply:
Thanks for the feedback, this makes sense.  I have it working with your suggestions, will push the changes now.